torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -497,6 +497,13 @@ Tensor sgn_backward(const Tensor& x, const Tensor& gx, const Tensor& sgn) {
   }
 }
 
+Tensor masked_fill_backward(const Tensor& grad, const Tensor& mask) {
+  // masked_select does not work well with functorch, as its shape is data-dependent
+  return areAnyTensorSubclassLike({grad, mask}) ?
+          at::where(mask, grad, 0).sum() :
+          grad.masked_select(mask).sum();
```

Comment:
That looks composite compliant to me. Why the if?

Reply:
As the comment says, the shape of the output of `masked_select` is data dependent and functorch errors out on this. Now, in my opinion, the batching rule for `masked_select` should not return a batch of vectors of different size, but a long vector (perhaps with batching size of 1 if functorch requires it). cc @zou3519

Reply:
fwiw, this is the error: https://github.com/pytorch/pytorch/runs/7859535004?check_suite_focus=true

Reply:
fwiw, this is the error: https://github.com/pytorch/pytorch/runs/7859535004?check_suite_focus=true

Reply:
> Now, in my opinion, the batching rule for masked_select should not return a batch of vectors of different size, but a long vector (perhaps with batching size of 1 if functorch requires it)

`vmap(f)(xs)` has semantics equivalent to the following: in the absence of side-effects, it is equivalent to running f in a for-loop on each x and stack-ing the results.

Because masked_select is guaranteed to return a 1D tensor, then vmap(masked_select) should return a 2D tensor

Reply:
Makes sense, thanks!

Reply:
minor nit for the comment: replace "functorch" with "vmap". The other functorch transforms don't have problems with masked_select, it's just vmap.